### PR TITLE
Forbid creation of non-faulting null-check nodes.

### LINKED ISSUE #44227: Forbid creation of non-faulting null-check nodes.
This is a follow-up for https://github.com/dotnet/runtime/pull/44059.

There are 2 methods that create `GT_NULLCHECK`: `gtNewNullCheck` and `gtChangeOperToNullCheck`. 
We have added an assert `assert(fgAddrCouldBeNull(addr))` to `gtNewNullCheck` and it could be profitable to add the same assert to `gtChangeOperToNullCheck` to delete unnecessary null-checks earlier and sometimes allow  the later phase to generate better code.

The difficulty is that `gtChangeOperToNullCheck` is used in different phases and in some of them it is not easy to delete an unnecessary node. An easier solution would be to replace the node with a nop and allow later phases to delete it also hit some asserts ([prototype ](https://github.com/sandreenko/runtime/commit/1dae992aedc2979b57dde3034e521bab3c211a16)),

for example in 
```
runtime\artifacts\tests\coreclr\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe "jitutils\bin\pmi.dll" PrepOne  runtime\artifacts\tests\coreclr\Windows_NT.x64.Release\Tests\Core_Root\mscorlib.dll 5480
```
we try to hoist such expression:
```
               [000048] n----+------                 |  +--*  OBJ       simd32<System.Numerics.Vector`1[Single]> // note that `n` in the flags means non-faulting.
               [000044] -----+-N----                 |  |  \--*  LCL_VAR   byref  V01 arg1
```
and we create
```
               [000092] ----G-------              *  COMMA     void
     (  7,  5) [000089] n---G--H----              +--*  OBJ       simd32<System.Numerics.Vector`1[Single]> <l:$341, c:$382>
     (  1,  1) [000090] -------N----              |  \--*  LCL_VAR   byref  V01 arg1         u:1 $81
               [000091] ------------              \--*  NOP       void
```
and call `gtExtractSideEffList` on `000089`, `gtExtractSideEffList` can't drop the tree because it is marked as `GTF_MAKE_CSE` so it will try to call `gtChangeOperToNullCheck` on an `OBJ` that can't fail.

The resulting tree with the current master is:
```
***** BB06
STMT00012 (IL   ???...  ???)
N004 (  2,  2) [000092] ----G-------              *  COMMA     void
N002 (  2,  2) [000089] ----G--H----              +--*  NULLCHECK int
N001 (  1,  1) [000090] -------N----              |  \--*  LCL_VAR   byref  V01 arg1         u:1 $81
N003 (  0,  0) [000091] ------------              \--*  NOP       void
```
it is not clear how would we like to transform it, because if the tree should be just deleted why did we choose this node for loop hoisting?

Note: I have run asm diffs with the prototype and have not seen asm diffs except for failing methods, so fixing this issue could be a clean-up only, rather than CQ, but it is not certain.

cc @AndyAyersMS 

category:cq
theme:null-checks

### PR DESCRIPTION
**Pull request by JulieLeeMSFT**
Fixes #44227.

There are 2 methods that create `GT_NULLCHECK`: `gtNewNullCheck` and `gtChangeOperToNullCheck`.
- https://github.com/dotnet/runtime/pull/44059 added `assert(fgAddrCouldBeNull(addr))` to `gtNewNullCheck` to avoid unnecessary null-checks early.
- This PR added an assert `assert(fgAddrCouldBeNull(tree->AsUnOp()->gtGetOp1()))` to `gtChangeOperToNullCheck` using @sandreenko's [prototype](https://github.com/sandreenko/runtime/commit/1dae992aedc2979b57dde3034e521bab3c211a16).

Edit: To delete an unnecessary node, we replaced the node with a `GT_NULLCHECK` or `GT_NOP` and allowed later phases to delete it.

### CODE REVIEW

**Comment by JulieLeeMSFT** [comment]
Fixed all asserts, but there are size and throughput regressions on Arm64. [Link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=64732&view=ms.vss-build-web.run-extensions-tab) 





@AndyAyersMS @kunalspathak PTAL.

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -7273,13 +7273,20 @@ void Lowering::LowerIndir(GenTreeIndir* ind)
 #if defined(TARGET_ARM64)
         // Verify containment safety before creating an LEA that must be contained.
         //
-        const bool isContainable = IsSafeToContainMem(ind, ind->Addr());
+        const bool isContainable = false;
+        if (ind->Addr() != nullptr)
+        {
+            const bool isContainable = IsSafeToContainMem(ind, ind->Addr());
```
Comment (markples) [inline src/coreclr/jit/lower.cpp:7279]: This is hiding the other `isContainable` and then going out of scope and the value is lost.

Comment (kunalspathak) [reply src/coreclr/jit/lower.cpp:7279]: Yep and this is the reason for so many diffs.

Comment (JulieLeeMSFT) [reply src/coreclr/jit/lower.cpp:7279]: oh I forgot to remove it when moving it out of if. Thanks for spotting it.

Comment (markples) [reply src/coreclr/jit/lower.cpp]: just a nit here - you could write it as

```
const bool isContainable = (ind->Addr() != nullptr) && IsSafeToContainMem(ind, ind->Addr());
```

if you wanted to preserve the `const`ness or shorten it.  I don't think it's a big deal.

Comment (JulieLeeMSFT) [reply src/coreclr/jit/lower.cpp]: Thanks for the code snippet. Will make that change.

**Comment by JulieLeeMSFT** [comment]
[Diff result](https://dev.azure.com/dnceng-public/public/_build/results?buildId=121090&view=ms.vss-build-web.run-extensions-tab):
- Small (~130 Bytes) code size improvements.
- Minor throughput diff (0.01%) as shown in the second screenshot below.
--------------------------------------------------------------------------------

image

image

**Comment by JulieLeeMSFT** [comment]
@dotnet/jit-contrib @BruceForstall, it is ready to review. 
Failure is a knonw issue: #75244.

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -8251,7 +8251,14 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                             // via an underlying address, just null check the address.
                             if (op1->OperIs(GT_FIELD, GT_IND, GT_OBJ))
                             {
-                                gtChangeOperToNullCheck(op1, block);
+                                if (fgAddrCouldBeNull(op1->AsUnOp()->gtGetOp1()))
```
Comment (BruceForstall) [inline src/coreclr/jit/importer.cpp:8254]: The `op1` (addr) of a static GT_FIELD can be `nullptr`, but `fgAddrCouldBeNull` doesn't check for that. Could we AV here?

File: src/coreclr/jit/lower.cpp [hunk 3]
```diff
@@ -7279,13 +7279,16 @@ void Lowering::LowerIndir(GenTreeIndir* ind)
...
         //
-        const bool isContainable = IsSafeToContainMem(ind, ind->Addr());
+        const bool isContainable = (ind->Addr() != nullptr) && IsSafeToContainMem(ind, ind->Addr());
 #else
         const bool isContainable         = true;
 #endif
 
-        TryCreateAddrMode(ind->Addr(), isContainable, ind);
-        ContainCheckIndir(ind);
+        if (!ind->OperIs(GT_NOP))
```
Comment (BruceForstall) [inline src/coreclr/jit/lower.cpp:7287]: The top of the function has:
```
assert(ind->OperIs(GT_IND, GT_NULLCHECK));
```
so isn't this `if` always true?

File: src/coreclr/jit/lower.cpp [hunk 4]
```diff
@@ -7279,13 +7279,16 @@ void Lowering::LowerIndir(GenTreeIndir* ind)
 #if defined(TARGET_ARM64)
         // Verify containment safety before creating an LEA that must be contained.
         //
-        const bool isContainable = IsSafeToContainMem(ind, ind->Addr());
+        const bool isContainable = (ind->Addr() != nullptr) && IsSafeToContainMem(ind, ind->Addr());
```
Comment (BruceForstall) [inline src/coreclr/jit/lower.cpp:7282]: Why is this necessary to check? Can GT_IND/GT_NULLCHECK have a null Addr?

File: src/coreclr/jit/lower.cpp [hunk 5]
```diff
@@ -7333,14 +7336,24 @@ void Lowering::TransformUnusedIndirection(GenTreeIndir* ind, Compiler* comp, Bas
...
+
     ind->ChangeType(comp->gtTypeForNullCheck(ind));
 
 #if defined(TARGET_ARM64) || defined(TARGET_LOONGARCH64)
     bool useNullCheck = true;
 #elif TARGET_ARM
     bool           useNullCheck          = false;
 #else  // TARGET_XARCH
-    bool useNullCheck = !ind->Addr()->isContained();
+    // bool useNullCheck = !ind->Addr()->isContained();
```
Comment (BruceForstall) [inline src/coreclr/jit/lower.cpp:7355]: Remove commented-out code.

Can you also please change:
```
#elif TARGET_ARM
```
to:
```
#elif defined(TARGET_ARM)
```

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/gentree.cpp [hunk 6]
```diff
@@ -16050,8 +16050,17 @@ void Compiler::gtExtractSideEffList(GenTree*     expr,
...
+                        if (m_compiler->fgAddrCouldBeNull(node->AsUnOp()->gtGetOp1()))
+                        {
+                            JITDUMP("Replace an unused OBJ/BLK node [%06d] with a NULLCHECK\n", dspTreeID(node));
+                            // It would change a non-faulting ind to a non-faulting nullcheck.
+                            m_compiler->gtChangeOperToNullCheck(node, m_compiler->compCurBB);
+                        }
+                        else
+                        {
+                            JITDUMP("Replace an unused OBJ/BLK node [%06d] with a NOTHING node\n", dspTreeID(node));
+                            node = m_compiler->gtNewNothingNode();
```
[the remark this answers is not in the document]
Comment (JulieLeeMSFT) [reply src/coreclr/jit/gentree.cpp:16062]: `Append(node)` sets class variable `GenTree* m_result`, and `node` is used either (1) to be directly assigned to `m_result` or (2) to make `comma` tree that is subsequently assigned to `m_result`. 
So, I think the code is correct.

File: src/coreclr/jit/compiler.cpp [hunk 7]
```diff
@@ -9953,6 +9953,7 @@ void Compiler::gtChangeOperToNullCheck(GenTree* tree, BasicBlock* block)
     assert(tree->OperIs(GT_FIELD, GT_IND, GT_OBJ, GT_BLK));
     tree->ChangeOper(GT_NULLCHECK);
     tree->ChangeType(gtTypeForNullCheck(tree));
+    assert(fgAddrCouldBeNull(tree->AsUnOp()->gtGetOp1()));
```
[the remark this answers is not in the document]
Comment (JulieLeeMSFT) [reply src/coreclr/jit/compiler.cpp:9956]: Added the line.

**Comment by JulieLeeMSFT** [comment]
@BruceForstall, it is ready to review.

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (JulieLeeMSFT) [reply src/coreclr/jit/importer.cpp]: Done.

File: src/coreclr/jit/gentree.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (JulieLeeMSFT) [reply src/coreclr/jit/gentree.cpp]: Done.

**Review by BruceForstall — APPROVED** [review]
